### ISSUE TITLE
Provide juliaw.exe to  suppress the terminal window on startup

### ISSUE THREAD

**Issue by mmoh** [JuliaLang/julia#13878]
Both python and java include a pthonw and javaw, version that supresses the terminal window on startup 

This summarizes why it's useful
"If you're going to call a python script from some other process (say, from the command line), use pythonw.exe. Otherwise, your user will continuously see a cmd window launching the python process. It'll still run your script just the same, but it won't intrude on the user experience.

An example might be sending an email; python.exe will pop up a CLI window, send the email, then close the window. It'll appear as a quick flash, and can be considered somewhat annoying. pythonw.exe avoids this, but still sends the email. " http://stackoverflow.com/questions/9705982/pythonw-exe-or-python-exe

If anyone has alternate suggestion on how to duplicate this functionality in julia that would also be helpful. 

This is likely only relevant for a Windows user.

**Comment by musmo** [comment]
Any tips on how one could go around doing this? I'm not an expert, but if there is some guide I could try?

**Comment by dhoegh** [comment]
My first thought was running:
```
spawn(detach(`julia -e "using IJulia; IJulia.notebook()"`))
```
But somehow the process do not get detached and hence the process dies when the parrent process is closed. It seems like the process is not detached correctly.

**Comment by tkelman** [comment]
Try changing https://github.com/JuliaLang/julia/blob/43c597bb3a24fae15e1c355897a6371362fc179a/ui/repl.c#L628 to `[w]WinMain` maybe?

**Comment by vtjnash** [comment]
there's a libuv option for hiding the window (https://github.com/libuv/libuv/blob/86051a49d8d0bb48d4372ce5c4caef83c5890c78/src/win/process.c#L1055-L1060). i think you may also need to provide stdio handles to spawn.

a more advanced solution that would require more design effort would involve manual handling of the terminal window: (ref http://stackoverflow.com/questions/13100785/mingw-build-gui-application-with-console)
- add `-mwindows` to the repl.c linking step in `ui/Makefile`
- allocate dummy stdio handles during julia's `init_stdio` (in `init.c`) since otherwise fd 0-2 will be invalid
- eventually call `AllocConsole` and `open("CONIN$")` to create the window on demand

**Comment by musmo** [comment]
It's been a while but I was able to play with this today and yesterday.

What seems to work is this add `-mwindows` to the repl.c and then compile

create a file `test.jlw` with
```
run(`notepad.exe`)
```
Then running with regular `julia.exe` without `-mwindows` creates a console for julia and opens a separate window for `notepad`. Using `-mwindows` works without creating a console.  (Windows has some confusion trying to assosciate  `.jlw`  to my compiled version of `julia`  since it has the same name as the precompiled version of julia 4 on my computer, but this is not important.)

@vtjnash I'm confused regarding your second bullet. What do you mean by "fd 0-2 will be invalid"
and also by
> eventually call AllocConsole and open("CONIN$") to create the window on demand
This would create another console which is what we are trying to avoid?

**Comment by vtjnash** [comment]
> I'm confused regarding your second bullet. What do you mean by "fd 0-2 will be invalid"
and also by

https://support.microsoft.com/en-us/kb/105305

If we link as `-mwindows`, it should be possible to have just one process (`julia.exe`), but provide a command line argument to tell Julia not to allocate a console (e.g. also `-mwindows`) – or just delay initialization until the program requests a console.

**Comment by PallHaraldsson** [comment]
>This is likely only relevant for a Windows user.

Yes, only Windows has GUI/window vs console distinction. The type (or some other possibilities) is encoded in the .EXE header.

>If we link as -mwindows, it should be possible to have just one process (julia.exe), but provide a command line argument to tell Julia not to allocate a console (e.g. also -mwindows)

I don't think that's possible, as I explained, but I would love to be proven wrong, and you link didn't work so it didn't help me...

When I, and most by now, run julia I actually I actually run juliaup launcher that starts Julia for me, with tiny startup overhead. I'm just thinking could it be there this juliaw.exe variant? I doubt it, also conversely would it interfere?

>there's a libuv option for hiding the window (https://github.com/libuv/libuv/blob/86051a49d8d0bb48d4372ce5c4caef83c5890c78/src/win/process.c#L1055-L1060). i think you may also need to provide stdio handles to spawn.

You mean the windows appears then is hidden and you might see it blinking very quickly? Than an imperfect workaround, though maybe ok.

>Both python and java include a pythonw.exe and javaw.exe that suppresses the terminal window on startup

For some reason, and I was always curious why they need it and we get away with not having the same. I think we simply don't have as good level of Windows (GUI) support.

The only way I can see we don't need this is if you fully compile with e.g. PackageCompiler.jl. I suppose it has the same limitation, but it (only) could provide those two options?

If we add juliaw.exe, then we likely need to add it for other platforms as a symlink, to just have tutorials working, even though it's a no-op variant there.

My current julia (i.e. the one juliaup starts) is only 23320 bytes (can maybe be made even smaller?), juliaw should be the same, or very similar. It loads more code that's much bigger, but I believe all of it would be shared, so this duplication isn't too costly so I think we should do this unless there's a good workaround I'm missing.
